internal/driver/writehandler.go

```diff
@@ -36,8 +39,44 @@ func (d *Driver) HandleWriteCommands(deviceName string, protocols map[string]mod
 		return err
 	}
 
+	endpoints, err := opcua.GetEndpoints(endpoint)
+	if err != nil {
+		d.Logger.Error("OPC GetEndpoints: %w", err)
+	}
+	credentials, err := getCredentials(d.serviceConfig.OPCUAServer.CredentialsPath)
+	if err != nil {
+		d.Logger.Error("getCredentials: %w", err)
+	}
+
+	username := credentials.Username
+	password := credentials.Password
+	policy := ua.SecurityPolicyURIBasic256Sha256
+	mode := ua.MessageSecurityModeSignAndEncrypt
+
+	ep := opcua.SelectEndpoint(endpoints, policy, mode)
+	c, err := generateCert() // This is where you generate the certificate
+	if err != nil {
+		d.Logger.Error("generateCert: %w", err)
+	}
+
+	pk, ok := c.PrivateKey.(*rsa.PrivateKey) // This is where you set the private key
+	if !ok {
+		log.Print("invalid private key")
+	}
+
+	cert := c.Certificate[0]
+
+	opts := []opcua.Option{
```

Comment:
This is repeating maybe we can create a function and reuse that.

Reply:
made some refactoring work, still potential but at least something for now :)